File: src/Rules/card.js
const joi = require("joi");

// ==================== 卡片基础 (card_base) ====================

// 创建卡片验证规则 - 根据 cardType 动态判断必填字段
exports.createCard = {
  body: joi.object({
    data: joi.object({
      // ========== 【通用必填字段】 ==========
      bankId: joi.string().required().messages({
        "any.required": "bankId 银行ID 不能为空",
      }),
      cardType: joi.string().valid("debit", "credit").required().messages({
        "any.only": "cardType 只能是 debit(借记卡) 或 credit(信用卡)",
        "any.required": "cardType 卡类型 不能为空",
      }),
      last4No: joi.string().required().messages({
        "any.required": "last4No 卡号后四位 不能为空",
      }),
      cardBin: joi.string().length(6).required().messages({
        "string.length": "cardBin 卡BIN前6位必须是6位数",
        "any.required": "cardBin 卡BIN前6位 不能为空",
      }),
      openDate: joi.string().required().messages({
        "any.required": "openDate 开卡日期 不能为空",
      }),
      expireDate: joi.string().required().messages({
        "any.required": "expireDate 过期日期 不能为空",
      }),
      sourceFrom: joi.string().required().messages({
        "any.required": "sourceFrom 数据来源 不能为空",
      }),
      // ========== 【信用卡必填字段】 ==========
      creditLimit: joi.number().when('cardType', {
        is: 'credit',
        then: joi.number().required().min(0).messages({
          "any.required": "creditLimit 信用额度（信用卡必填）不能为空",
          "number.min": "信用额度不能为负数",
        }),
        otherwise: joi.number().default(0),
      }),
      tempLimit: joi.number().default(0),
      billDay: joi.number().when('cardType', {
        is: 'credit',
        then: joi.number().required().messages({
          "any.required": "billDay 账单日（信用卡必填）不能为空",
        }),
        otherwise: joi.number().default(0),
      }),
      repayDay: joi.number().when('cardType', {
        is: 'credit',
        then: joi.number().required().messages({
          "any.required": "repayDay 还款日（信用卡必填）不能为空",
        }),
        otherwise: joi.number().default(0),
      }),
      annualFee: joi.number().when('cardType', {
        is: 'credit',
        then: joi.number().required().messages({
          "any.required": "annualFee 年费（信用卡必填）不能为空",
        }),
        otherwise: joi.number().default(0),
      }),
      feeFreeRule: joi.string().when('cardType', {
        is: 'credit',
        then: joi.string().required().messages({
          "any.required": "feeFreeRule 免年费规则（信用卡必填）不能为空",
        }),
        otherwise: joi.string().allow('', null).default(''),
      }),
      // ========== 【通用非必填字段】 ==========
      cardLevel: joi.string().default('普卡'),
      mainSub: joi.string().default('主卡'),
      cardOrg: joi.string().default('银联'),
      cardLength: joi.string().default('19'),
      alias: joi.string().allow('', null).default(''),
      cardImg: joi.string().allow('', null).default(''),
      currency: joi.string().default('CNY'),
      status: joi.string().default('正常'),
      isDefault: joi.boolean().default(false),
      isHide: joi.boolean().default(false),
      sort: joi.number().default(99),
      tag: joi.string().allow('', null).default(''),
      remark: joi.string().allow('', null).default(''),
      color: joi.string().default('#0052cc'),
    }).unknown(true),
  }).unknown(true),
};

// 更新卡片验证规则（非必填）
exports.updateCard = {
  body: joi.object({
    data: joi.object({
      bankId: joi.string().allow('', null),
      cardType: joi.string().valid("debit", "credit"),
      cardLevel: joi.string().allow('', null),
      mainSub: joi.string().allow('', null),
      cardOrg: joi.string().allow('', null),
      cardLength: joi.string().allow('', null),
      last4No: joi.string().allow('', null),
      cardBin: joi.string().allow('', null),
      alias: joi.string().allow('', null),
      cardImg: joi.string().allow('', null),
      openDate: joi.string().allow('', null),
      expireDate: joi.string().allow('', null),
      billDay: joi.number().allow(null),
      repayDay: joi.number().allow(null),
      currency: joi.string().allow('', null),
      status: joi.string().allow('', null),
      isDefault: joi.boolean(),
      isHide: joi.boolean(),
      sort: joi.number(),
      tag: joi.string().allow('', null),
      remark: joi.string().allow('', null),
      color: joi.string().allow('', null),
      annualFee: joi.number().allow(null),
      feeFreeRule: joi.string().allow('', null),
      sourceFrom: joi.string().allow('', null),
    }).unknown(true),
  }).unknown(true),
};

// ==================== 卡片账单 (card_bill) ====================

// 创建账单验证规则
exports.createBill = {
  body: joi.object({
    data: joi.object({
      // 【前端必填】
      cardId: joi.string().required().messages({
        "any.required": "cardId 卡片ID 不能为空",
      }),
      creditLimit: joi.number().required().messages({
        "any.required": "creditLimit 信用额度 不能为空",
      }),
      tempLimit: joi.number().default(0),
      pointsRate: joi.number().default(1),
      // 【可选】
      billMonth: joi.string().pattern(/^\d{4}-\d{2}$/).messages({
        "string.pattern.base": "billMonth 格式应为 YYYY-MM",
      }),
      // 【后端计算/默认值】
      remindSwitch: joi.boolean().default(true),
      remindDays: joi.number().default(3),
    }).unknown(true),
  }).unknown(true),
};

// 更新账单验证规则
exports.updateBill = {
  body: joi.object({
    data: joi.object({
      creditLimit: joi.number(),
      tempLimit: joi.number(),
      pointsRate: joi.number(),
      repaid: joi.number(),
      remindSwitch: joi.boolean(),
      remindDays: joi.number(),
    }).unknown(true),
  }).unknown(true),
};

// ==================== 还款记录 (card_repay) ====================

// 创建还款记录验证规则
exports.createRepay = {
  body: joi.object({
    data: joi.object({
      cardId: joi.string().required().messages({
        "any.required": "cardId 卡片ID 不能为空",
      }),
      billId: joi.string().required().messages({
        "any.required": "billId 账单ID 不能为空",
      }),
      repayAmount: joi.number().required().min(0.01).messages({
        "any.required": "repayAmount 还款金额 不能为空",
        "number.min": "还款金额必须大于 0",
      }),
      repayMethod: joi.string().required().valid('card', 'balance', 'bank_card', 'cash').messages({
        "any.required": "repayMethod 还款方式 不能为空",
        "any.only": "repayMethod 只支持: card/balance/bank_card/cash",
      }),
      repayMethodCardId: joi.string().when('repayMethod', {
        is: 'bank_card',
        then: joi.string().required().messages({
          "any.required": "银行卡还款时 repayMethodCardId 不能为空",
        }),
        otherwise: joi.string().allow('').optional(),
      }),
      repayTime: joi.string().required().pattern(/^\d{4}-\d{2}-\d{2}$/).messages({
        "any.required": "repayTime 还款时间 不能为空",
        "string.pattern.base": "repayTime 格式应为 YYYY-MM-DD",
      }),
      remark: joi.string().allow('').optional(),
    }).unknown(true),
  }).unknown(true),
};

// 更新还款记录验证规则
exports.updateRepay = {
  body: joi.object({
    data: joi.object({
      repayAmount: joi.number(),
      repayMethod: joi.string(),
      repayTime: joi.string(),
      remark: joi.string(),
    }).unknown(true),
  }).unknown(true),
};

File: app/src/main/java/com/live/finance/data/model/Card.kt
package com.live.finance.data.model

/**
 * 银行卡/账户（对应 GET /card 行；后端 LEFT JOIN bus_category(type=bank) 带出 bank_name）。
 *
 * 注意（后端 findAll 的字段裁剪，见 api/src/modules/card/model/index.js）：
 * - cardType=credit 时返回：card_bin/card_length/last4_no/alias/open_date/expire_date/card_org/tag/
 *   bill_day/repay_day/credit_limit/temp_limit/points_rate/share_pool_id/card_img/color + bank_name；
 * - cardType=debit 时返回：card_bin/card_length/last4_no/alias/open_date/expire_date/card_org/tag/
 *   card_img/color + bank_name（无额度/账单日等字段）；
 * - 未筛选时按 credit 字段集返回。
 * 故所有「未返回」字段一律走默认值，UI 需容忍空值。
 */
data class Card(
    val id: String,
    val cardType: String = "debit",          // debit / credit（另有 virtual_cash=xxxx / virtual_balance=yyyy）
    val bankId: String = "",
    val bankName: String = "",
    val alias: String = "",
    val last4: String = "",
    val cardBin: String = "",
    val cardOrg: String = "",
    val cardLevel: String = "",
    val mainSub: String = "主卡",
    val cardLength: String = "19",
    val cardImg: String = "",
    val color: String = "",
    val currency: String = "CNY",
    val status: String = "正常",
    val isDefault: Boolean = false,
    val isHide: Boolean = false,
    /** 排序位（1-N；后端 findAll 未 SELECT 此列，列表响应通常为空，靠 sort 接口写入）。 */
    val sort: Int = 99,
    val tag: String = "",
    val remark: String = "",
    val sourceFrom: String = "",
    val openDate: String = "",
    val expireDate: String = "",
    // ===== 信用卡专属 =====
    val billDay: Int = 0,
    val repayDay: Int = 0,
    val creditLimit: Double = 0.0,
    val tempLimit: Double = 0.0,
    val pointsRate: Double = 1.0,
    val sharePoolId: String = "",
    val annualFee: Double = 0.0,
    val feeFreeRule: String = "",
) {
    val isCredit: Boolean get() = cardType == "credit"
    val isVirtual: Boolean get() = cardType.startsWith("virtual_")

    /** 「别名 || 银行名(尾号xxxx)」——对齐 web `alias || 银行名(尾号) ` 兜底。 */
    val displayName: String
        get() = alias.ifBlank {
            val b = bankName.ifBlank { "未命名卡" }
            if (last4.isNotBlank()) "$b($last4)" else b
        }

    /** 卡面主色（后端 color 形如 `#0052cc`），空则给主题蓝兜底。 */
    val displayColor: String get() = color.ifBlank { "#0052cc" }

    /**
     * 脱敏卡号（对齐 web CardStack.formatCardNo）：BIN 之后补 `*` 至 (cardLength-4)，末尾接后 4 位。
     * 例：cardBin=622202 cardLength=19 last4=0088 → `622202***********0088`（13 个星）。
     */
    val maskedNo: String
        get() {
            if (last4.isBlank() && cardBin.isBlank()) return "****"
            val total = cardLength.toIntOrNull()?.takeIf { it in 8..25 } ?: 19
            val stars = (total - cardBin.length - last4.length).coerceAtLeast(0)
            return cardBin + "*".repeat(stars) + last4
        }

    /** 按每 4 位分组（web 卡面展示用 `6222 0212 **** **** 0088` 风格）。 */
    val groupedNo: String
        get() = maskedNo.chunked(4).joinToString(" ")
}
